16.0 mig customer manager base

Honeyxilia opened the pull request.

fcayre committed:
Add an empty customer_manager_base module

fcayre committed:
Add groups and rules to allow fine-tuning sale, accounting and project documents access per portal user

fcayre committed:
Add action and wizard for sale managers to set management groups on customer partners

fcayre committed:
customer_manager_base i18n update

fcayre committed:
Add ability to configure portal access to project tasks depending on group memberships

Portal users may access tasks they do not directly follow but one of their colleagues,
if they belong to one of the groups selected in the project's
portal_visibility_restriction_group_ids field.

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Fix invalid ir.rule domain expression for project task

github-grap-bot committed:
[BOT] post-merge updates

fcayre committed:
Remove now useless wizard to assign groups to customer partners

It is now replaced by the customer_team_manager module.

github-grap-bot committed:
[BOT] post-merge updates

Honeyxilia committed:
[IMP] customer_manager_base: black, isort, prettier

Honeyxilia committed:
[MIG] customer_manager_base: Migration to 16.0

Honeyxilia committed:
Updated account.invoice related fields to account.move fields

Honeyxilia committed:
Replaced SavepointCase with TransactionCase

Honeyxilia committed:
Replaced model relation tuples by Command methods

In other modules, it seems that older relation tuples to create model relations have been replaced by the methods of the Command class - so, I changed them here

Honeyxilia committed:
Changed _give_portal_access behavior to match changes in portal.wizard

- The `in_portal` variable was changed into `is_portal` (and a new `is_internal` variable has been addded)
- `action_apply` no longer exists, so we use the new `action_grant_access` method

Honeyxilia committed:
Changed test_method name and comment to clarify purpose

Since this method tests if a user have access to a task, but with varying conditions (owning for invoices/moves, and follow for tasks and sales), this modification allows to clarify what the method actually tests

Honeyxilia committed:
Fixed sale.order related tests

The test partner were changed from the base sale.order partner since it's subscribed by default in the sale message followers - so we copy the sale partner, and change the parent to be the sale partner

Honeyxilia committed:
Changed outdated demo account.move record

The newly used record is generated and added to the demo pool in the indicated file, in a demo data generator

Honeyxilia committed:
Changed message subscribing method

Honeyxilia committed:
Moved test object creations in setUpClass method

I kept setUp classes, because the _grant_portal_access method uses assert methods, which are instance methods and not class methods - as such, they cannot be used within the setUpClass method as they are

9184Lucky commented on customer_manager_base/security/rules.xml:
```diff
@@ -29,25 +29,25 @@
-    <field name="name">Portal: see company invoices if accounting customer</field>
-    <field name="model_id" ref="account.model_account_invoice" />
+  <record model="ir.rule" id="account_move_rule_customer_accounting">
+    <field name="name">Portal: see company moves if accounting customer</field>
```
ici j'aurais gardé "invoices"

Honeyxilia replied:
j'ai rétabli l'ancien texte c:

9184Lucky commented on customer_manager_base/tests/test_ir_rules.py:
```diff
@@ -14,8 +14,9 @@ def seen(self, entities, user):
         portal_wizard = model.sudo().create({})
-        portal_wizard.user_ids.update({"in_portal": True})
-        portal_wizard.action_apply()
+        for wizard_user in portal_wizard.user_ids:
```
Mmmh y'a moyen d'écrire ça de manière plus efficace à mon avis (qui évite de boucler sur les user_ids) en fesant un truc du genre:
portal_wizard.user_ids.filtered_domain([("is_portal", "=", False)]).action_grant_access()

En plus en regardant la définition de la fonction action_grant_access j'ai l'impression que la version actuelle plante si y'a plusieurs users dans user_ids

Honeyxilia replied:
J'ai intégré ta recommandation, ceci améliorera en plus la couverture du code à 100% c:

Concernant la présence de plusieurs users dans user_ids, ce n'est pas problématique ici, car s'il n'y a qu'un seul partner dans le contexte, alors ceci ne créera qu'au maximum 1 user_id (cf. `_default_partner_id`)

Par contre, cela vaudrait le cout de bien mettre un `partner.ensure_one()` au début de la méthode

9184Lucky commented on customer_manager_base/tests/test_ir_rules.py:
```diff
@@ -68,10 +68,17 @@ class PortalInvoiceIrRulesTC(PortalIrRulesTC, TransactionCase):
     @classmethod
     def setUpClass(cls):
         super().setUpClass()
+        test_company = cls.env.ref("l10n_fr.demo_company_fr", False)
```
A mon avis y'a moyen de faire plus proprement ici (cf le message de Flo et la PR 391). Si besoin on peut regarder ensemble.

Honeyxilia replied:
(Vu en visioconférence, et modification poussée)

9184Lucky reviewed (commented):
Quelques commentaires mais sinon c'est bon pour moi !

Honeyxilia committed:
Fixed invoice test data for production tests + Added coverage flags

At the start of the setUpClass method in PortalInvoiceIrRulesTC, we check if the environnement company has no chart template, we load accounts for the environnement company and reinstall demo data so it corresponds with the current company

fcayre commented:
/ocabot merge patch
